pkg/reconciler/testing/reconcile.go

```diff
@@ -0,0 +1,392 @@
+			},
+		},
+
+		// Lifecycle
+
+		{
+			Name: "Adapter becomes Ready",
+			Key:  tKey,
+			Objects: []runtime.Object{
+				newAdressable(),
+				newEventSource(withSink, notDeployed),
+				newAdapterDeployment(ready),
+			},
+			WantStatusUpdates: []clientgotesting.UpdateActionImpl{{
+				Object: newEventSource(withSink, deployed),
+			}},
+		},
+		{
+			Name: "Adapter becomes NotReady",
+			Key:  tKey,
+			Objects: []runtime.Object{
+				newAdressable(),
+				newEventSource(withSink, deployed),
+				newAdapterDeployment(notReady),
+			},
+			WantStatusUpdates: []clientgotesting.UpdateActionImpl{{
+				Object: newEventSource(withSink, notDeployed),
+			}},
+		},
+		{
+			Name: "Adapter is outdated",
+			Key:  tKey,
+			Objects: []runtime.Object{
+				newAdressable(),
+				newEventSource(withSink, deployed),
+				newAdapterDeployment(ready, bumpImage),
+			},
+			WantUpdates: []clientgotesting.UpdateActionImpl{{
+				Object: newAdapterDeployment(ready),
+			}},
+			WantEvents: []string{
+				updateAdapterEvent(n),
+			},
+		},
+
+		// Errors
+
+		{
+			Name: "Sink goes missing",
+			Key:  tKey,
+			Objects: []runtime.Object{
+				/* sink omitted */
+				newEventSource(withSink, deployed),
+				newAdapterDeployment(ready),
+			},
+			WantStatusUpdates: []clientgotesting.UpdateActionImpl{{
+				Object: newEventSource(withoutSink, deployed),
+			}},
+			WantEvents: []string{
+				badSinkEvent(),
+			},
+			WantErr: true,
+		},
+		{
+			Name: "Fail to create adapter deployment",
+			Key:  tKey,
+			WithReactors: []clientgotesting.ReactionFunc{
+				rt.InduceFailure("create", "deployments"),
+			},
+			Objects: []runtime.Object{
+				newAdressable(),
+				newEventSource(withSink),
+			},
+			WantCreates: []runtime.Object{
+				newAdapterDeployment(),
+			},
+			WantStatusUpdates: []clientgotesting.UpdateActionImpl{{
+				Object: newEventSource(unknownDeployedWithError, withSink),
+			}},
+			WantEvents: []string{
+				failCreateAdapterEvent(n),
+			},
+			WantErr: true,
+		},
+		{
+			Name: "Fail to update adapter deployment",
+			Key:  tKey,
+			WithReactors: []clientgotesting.ReactionFunc{
+				rt.InduceFailure("update", "deployments"),
+			},
+			Objects: []runtime.Object{
+				newAdressable(),
+				newEventSource(deployed, withSink),
+				newAdapterDeployment(ready, bumpImage),
+			},
+			WantUpdates: []clientgotesting.UpdateActionImpl{{
+				Object: newAdapterDeployment(ready),
+			}},
+			WantEvents: []string{
+				failUpdateAdapterEvent(n),
+			},
+			WantErr: true,
+		},
+
+		// Edge cases
+
+		{
+			Name:    "Reconcile a non-existing object",
+			Key:     tKey,
+			Objects: nil,
+			WantErr: false,
+		},
+	}
+
+	testCases.Test(t, MakeFactory(ctor))
+}
+
+/* Event sources */
+
+// eventSourceCtor creates a copy of the given source and returns a function
+// that can apply options to that sources.
+func eventSourceCtor(src v1alpha1.AWSEventSource, arn arn.ARN,
+	eventTypes []string) func(...sourceOption) v1alpha1.AWSEventSource {
```

Comment:
I have an idea to simplify this a bit.

The source contains everything that's needed to get the ARN and event types, so I might as well pass just the source and add a new `GetEventTypes()` method to the AWSEventSource interface method to fetch the event types. I just wanted to avoid touching the interface in this PR.

Reply:
Done